Fixes #26020 - report templates in background process

ezr-ondrej opened the pull request:
Moves the report templates rendering in background.
Report output is now stored in temporary file, the one with database storage follows.

Depends on https://github.com/Dynflow/dynflow/pull/317 merge and dynflow bump.

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

ezr-ondrej committed:
Fixes #26020 - report templates in background process

ohadlevy commented on lib/foreman/stored_result.rb:
```diff
@@ -0,0 +1,43 @@
+    class TmpFileProvider
+      DIRECTORY = Rails.root.join('tmp')
+
+      def write(result_key, result)
```
IMHO this is wrong, we should not support writing to files as this can break deployment that scale across multiple hosts, did you consider using the rails cache instead?

timogoebel replied:
I concur. Writing to files is introducing technical debt. What's wrong with streaming the data? Why do we have to do it async?

ezr-ondrej replied:
This is another point to be hopefully opinionated :)
Rails.cache is not really persistent across instances eighter (you can change the implementation from database to file), files are than the same.
I have introduced this class, basically just to make an API to the StoredResult.
Now we can have these tree providers and I would like to end up at one:
1. Rails.cache
2. TempFile
3. Database - see `stored_results`

ezr-ondrej replied:
@timogoebel there is big concern about the reports taking too much of a time to generate.
In my opinion "too much" would be one minute and more ( you are not really happy to wait for it, so you will let that to be sent to your inbox, what is going to be next PR of mine ).

timogoebel replied:
I just saw a talk at a ruby conference where somebody actually streamed the data directly from the database to the client. (csv genereated in sql, stream from postgresql via actioncontroller::metal to the client).

ohadlevy replied:
> This is another point to be hopefully opinionated :)
What do you mean?

> Rails.cache is not really persistent across instances eighter (you can change the implementation from database to file), files are than the same.
Users with multi host behind a LB already knows that, and use a plugin such as foreman memcache to solve this issue.

> I have introduced this class, basically just to make an API to the StoredResult.
> Now we can have these tree providers and I would like to end up at one:
> 
> 1. Rails.cache
> 2. TempFile
> 3. Database - see `stored_results`

I would argue we can do 1 or 3, technically, 2 will work (we already do it in the smart proxy to store scap reports), but I think for foreman ability to scale we should not introduce it (e.g. it would require shared storage etc)
1. already offer abstraction between memory/file/shared service, and if your data is temporary, imho its a good solution, otherwise, consider using a metal controller to store the data as blob in your db.

timogoebel replied:
S3 compatible storage would be perfect for this, actually.

ohadlevy replied:
Agreed :) or even if we just be running on kubernetes... :)

ezr-ondrej replied:
@timogoebel well, it is not a problem to stream from ruby,  but to receive on the client without taking unecessary time from web worker. Even though it would be actually nice to think about the cases, when the result is not fitting to the memory (but it is a real corner case because there is a problem to process such reports on client anyway). And the most spent time is expected for rendering the report, the streaming should be relatively fast compared to the rendering.

ezr-ondrej replied:
@ohadlevy well it would makes sense actually, but the problem is, that you are choosing the storage for cache for serving like a cache the best and you are not changing it for reports, so it can make sense for cache to be in the file, even though you got more instances, because you don't care that the content is processed on every worker, but this would make the polling to fail, if you are hitting the other instance, than it was processed on.

S3 compatible it is very simple to do actully, but not really good like a default option 🤣
(well you can make some reports, but buy and configure S3 first for them...)

ezr-ondrej replied:
For me the best option would be to give a user choice between streaming through websockets or send by email. In that case you don't actually need any storage.

But we don't have websockets 😢

timogoebel replied:
Would you send a link to the file via mail or attach it to the mail?

ezr-ondrej replied:
@timogoebel if it would be meant to avoid storing the results, than sending right away, so I didn't thought about sending link, but if we are storing the results already, we can give that choice actually, but the stored results are not meant to be available for more than 24 hours anyway :thinking:

ezr-ondrej replied:
Ok, guys, I have chosen the database, with addition, that in the future, it would be nice to have an pulp storage option, if there happens to be one alongside the foreman.

timogoebel commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -107,12 +107,22 @@ def export
         if @composer.valid?
-          response = @report_template.render(params: params, template_input_values: @composer.template_input_values)
-          send_data response, :filename => @report_template.suggested_report_name.to_s
+          if params[:foreground]
```
Isn't this introducing an incompatible api change?

ezr-ondrej replied:
It is, @ares asked me to bring the discussion on, if we would want to make the generation on backgroud a default option.

ezr-ondrej replied:
I have made new endpoint for asynchronous generations, so we don't break the compatibility and actually leave it up to the user educated decision to use the right one for his usecase :)

timogoebel commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -107,12 +107,22 @@ def export
+      param :foreground, :bool, allow_nil: true
+      description <<-DOC
+        The reports are generated by background job. It returns an url to poll for data
+        The <b>foreground</b> parameter is used for to force report to be generated in foreground.
```
nit: `is used to force the report to be returned synchronously`. `in foreground` is the way it's implemented, not what the user sees when he does an api request.

ezr-ondrej replied:
I have rewritten the endpoints afterall.

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -104,15 +104,16 @@ def export
+      api :POST, "/report_templates/:id/generate/", N_("Generate a report from template")
       param :id, :identifier, :required => true
       param :input_values, Hash, :desc => N_('Hash of input values where key is the name of input, value is the value for this input')
+      param :gzip, :bool, desc: N_('Gzips the response'), default_value: false
```
`Compress the report uzing gzip` would sound slightly better I think

ezr-ondrej replied:
True, thanks :+1:

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -121,6 +122,48 @@ def generate
+      api :GET, "/report_templates/:id/report_data/:job_id", N_("Return a generated report")
+      param :id, :identifier, required: true
+      param :job_id, String, required: true
+      param :gzip, :bool, desc: N_('Gzips the response'), default_value: false
```
the same string as above

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -129,12 +172,24 @@ def action_permission
+      end
+
+      def send_report_data(template, data, gzip = false)
+        if gzip
```
If the report is big, we'd gzip as part of the web process again, I think we should do that in the background task. I understand this gives us possibility to download the same report either gzipped or not, but I think that's fine.

ohadlevy replied:
does streaming matters (similar to csv?)

ezr-ondrej replied:
well the send_data is already streamed response, so only improovement there is to achieve, is to stream data in chunks. Which would make sense for huge files, but there would be problem with the database storage limit.

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -129,12 +172,24 @@ def action_permission
+        Rails.application.dynflow.world.persistence.load_execution_plan(plan_id)
+      end
+
+      def send_report_data(template, data, gzip = false)
```
it seems we no longer set 
```
headers["Cache-Control"] = "no-cache"
headers["Content-Disposition"] = %(attachment; filename="#{@template.suggested_report_name}")
```

ezr-ondrej replied:
Well, I don't understand really well what it have been doing there anyway :thinking: Content-Disposition header is set by send_data, cache-control is set by send_data to `private` and it should be ok.

ares commented on app/jobs/template_render_job.rb:
```diff
@@ -0,0 +1,10 @@
+  queue_as :default
+
+  def perform(composer_params)
+    User.as_anonymous_admin do
```
This does not seem as a good idea, today the load_* marcros respect current user. We probably need to pass the User.current.id here and then "impersonate" such user. Or am I missing something?

ezr-ondrej replied:
yeah, I have send the user login, because that is used to impersonalizade user in the end. But not sure if it is a good idea...

ezr-ondrej replied:
user_id is used now

ares commented on app/models/stored_result.rb:
```diff
@@ -0,0 +1,11 @@
+class StoredResult < ApplicationRecord
+  def self.write(result_key, result_value)
+    record = self.find_or_create_by(key: result_key)
+    record.update(value: result_value)
```
I think it would be nicer to split this to two cases, if it exists then, else create with key and result_value. The difference is that we don't do save of new record in 2 queries.

ezr-ondrej replied:
Yeah, I think this one is bit better anyway - more clear (`self.find_by(key: result_key) || new(key: result_key)`)

ares commented on app/models/stored_result.rb:
```diff
@@ -0,0 +1,11 @@
+class StoredResult < ApplicationRecord
+  def self.write(result_key, result_value)
```
should it also accept `expire_at` info?

ezr-ondrej replied:
Yeah, it should for sure, thanks. It is defaulting to 24 hours.

ares commented on app/models/stored_result.rb:
```diff
@@ -0,0 +1,11 @@
+class StoredResult < ApplicationRecord
```
Since this is kind of generic store that we can use elsewhere, should the name be a bit more generic? E.g. StoredRecord, StoredValue, DataBucket. I can imagine using it from storing user preferences, things like user cancelled tour for a given page etc. So the Result might be a bit misleading.

ezr-ondrej replied:
Renamed to StoredValue, it seems as more generic name :) (not sure if not too generic, but we will see.

ares reviewed (changes requested):
I know it's WIP but wanted to provide a feedback ASAP. I think this goes the right direction. The only store we'll now support is DB which addresses HA concerns. If we decide to use something different on backed, the small StoredResult can be adapted. I'd appreciate comments on this from @ohadlevy and @timogoebel who raised the concern.

@amirfefer could you also take a look at the JS code please?

I did a quick testing (quick just because this is WIP so far). When I click submit button, there's no indication I should be waiting for report to be rendered. If nothing else, we should display the spinner and stop it when the file is ready. As we discussed it would be also nice to display the link for downloading the report if I no longer want to wait on this page. I'm not sure if this is a regression from current behavior, but would be good to address here to. Also the message above the rendering form should be updated, explaining how the rendering process work now.

Regarding authorization (both API and UI), downloading of stored results should be authorized with 'generate' permission and only the user who generated it (or admin) should be able to download it. Which I think becomes more problematic, because we don't store the user anywhere.

Good progress so far, looking forward for more updates!

ohadlevy commented on app/views/report_templates/generate.html.erb:
```diff
@@ -31,3 +31,21 @@
   <%= submit_or_cancel f, false, { data: { disable_with: false } } %>
 <% end %>
+
+<script type="text/javascript">
```
why is this part of the view vs under webpack?

ares replied:
I also wonder if we could move it out :-)

ezr-ondrej replied:
I've contained it in the webpack file initialize and just calling its function.

ohadlevy commented on webpack/assets/javascripts/react_app/redux/actions/templates/index.js:
```diff
@@ -0,0 +1,39 @@
+
+const defaultPollingInterval = 2500;
+const pollingInterval =
+  process.env.NOTIFICATIONS_POLLING || defaultPollingInterval;
```
this seems like a copy/paste, I wonder if we can generalize it somehow

ezr-ondrej replied:
I don't want to introduce another env variable just for polling, so I have reused it. It can be named differently, but we would have to rename it, and it is bit intrusive change :thinking:

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -7,7 +7,8 @@ class ReportTemplatesController < V2::BaseController
       before_action :find_optional_nested_object
-      before_action :find_resource, :only => %w{show update destroy clone export generate}
+      before_action :find_resource, :only => %w{show update destroy clone export generate schedule_report report_data}
+      # before_action :check_data_permissions, only: 'report_data'
```
leftover?

ezr-ondrej replied:
yeah, but I have used it in the end (viz `load_and_authorize_plan` before action)

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -121,6 +123,48 @@ def generate
+      def schedule_report
+        @composer = ReportComposer.from_api_params(params)
+        if @composer.valid?
+          job = TemplateRenderJob.perform_later(@composer.to_param, as_user: User.current.login, gzip: params[:gzip])
```
would it be safer to use `User.current.id`? numbers are safer for serialization also login can change in time

ezr-ondrej replied:
probably, seems safer, right? :laughing:

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -129,12 +173,30 @@ def action_permission
+
+      def send_plan_data(plan)
+        composer_attrs, options = plan_arguments(plan)
+        if User.current.admin? || options['as_user'] == User.current.login
```
this should use user id ideally

ezr-ondrej replied:
done :)

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -129,12 +173,30 @@ def action_permission
+      def send_plan_data(plan)
+        composer_attrs, options = plan_arguments(plan)
+        if User.current.admin? || options['as_user'] == User.current.login
+          composer = ReportComposer.new(composer_attrs, gzip: options['gzip'])
```
I suppose this is here just so we can get the suggested mime_type and filename, correct?

ezr-ondrej replied:
and to wrap value reading, but yeah :dart: you think it is useless to do?

ares replied:
no that's fine, just wanted to make sure I read it correctly :-)

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -129,12 +173,30 @@ def action_permission
+          composer = ReportComposer.new(composer_attrs, gzip: options['gzip'])
+          send_data composer.stored_result(plan.id), type: composer.mime_type, filename: composer.report_filename
+        else
+          render_error('access_denied', :status => :unauthorized, :locals => { :details => _('Admin permissions required') })
```
I think there's a better helper `deny_access` that should be used here, see https://github.com/theforeman/foreman/blob/develop/app/controllers/api/base_controller.rb#L204-L207

ezr-ondrej replied:
Yeah, I have rewritten the authorization to be before_action

ares commented on app/jobs/template_render_job.rb:
```diff
@@ -0,0 +1,11 @@
+  queue_as :default
+
+  def perform(composer_params, opts = {})
+    user_login = opts[:as_user]
```
if we use user id, here's is the place we'd need to find user by id and load his/her login

ezr-ondrej replied:
Done

ares commented on app/models/report_composer.rb:
```diff
@@ -57,7 +60,8 @@ def convert_input_names_to_ids(template_id, input_values)
   end
 
-  def initialize(params)
+  def initialize(params, gzip: false)
```
if we ever expose this as UI parameter too, it would make more sense to be in params, how about we don't use keyword here but rather

`@gziped_render = !!params[:gzip]`

like we do in API

ares commented on app/models/report_composer.rb:
```diff
@@ -112,4 +120,33 @@ def input_value_for(input)
+
+  def render(mode: Foreman::Renderer::REAL_MODE, gzip: false, **params)
+    result = @template.render(mode: mode, template_input_values: template_input_values, **params)
+    result = ActiveSupport::Gzip.compress(result) if gzip
```
why we don't use `gzip?` here? it's already set in initializer

ares commented on app/models/stored_value.rb:
```diff
@@ -0,0 +1,23 @@
+##
+# This class represents temporary storage for values processed in background.
+# This is a database solution, wich should not be the only one, as it has many limitations.
+
```
I don't like empty lines between docs and the code, I think at least for consistency reasons, we shouldn't do it :-)

ares commented on app/models/stored_value.rb:
```diff
@@ -0,0 +1,23 @@
+require 'base64'
+
+##
```
we usually don't format docs like this

ares commented on app/models/stored_value.rb:
```diff
@@ -0,0 +1,23 @@
+  def self.write(result_key, result_value, expire_at: nil)
+    record = self.find_by(key: result_key) || new(key: result_key)
+    record.value = Base64.encode64(result_value)
+    record.expire_at = expire_at || Time.now + 1.day
```
Now I wonder if we don't limit us too much here, what if we want to use the storage for something that never expires? I like the fact it takes expire_at as argument, if we set `Time.now + 1.day` from place we actually write to store, it would be imho better. The expiration interval is probably always specific to what we store.

ezr-ondrej replied:
Ok, but it allows permanent values! But done that :)

ares commented on app/models/stored_value.rb:
```diff
@@ -0,0 +1,23 @@
+
+  def self.write(result_key, result_value, expire_at: nil)
+    record = self.find_by(key: result_key) || new(key: result_key)
+    record.value = Base64.encode64(result_value)
```
is base64 required? how about we use some blob type in DB? encoding and decoding of 16 MB seems unnecessary

ezr-ondrej replied:
Moved to binary. Do not know, why I came up with this instead of binary in DB :(

ares commented on webpack/assets/javascripts/foreman_template_inputs.js:
```diff
@@ -1,6 +1,10 @@
 import $ from 'jquery';
 
+import store from './react_app/redux';
+
```
nit: I see blank lines between imports, in file below there are no blank lines in between, should we just drop them here?

ezr-ondrej replied:
felt better at the time of writing. Deleted that one.

ares commented on app/jobs/template_render_job.rb:
```diff
@@ -0,0 +1,11 @@
+    user_login = opts[:as_user]
+    User.as user_login do
+      composer = ReportComposer.new(composer_params, gzip: opts[:gzip])
+      composer.render_to_store(provider_job_id)
```
provider_job_id I suppose gives us dynflow task id, correct?

ezr-ondrej replied:
:dart:

ezr-ondrej commented on webpack/assets/javascripts/react_app/common/testHelpers.js:
```diff
@@ -58,6 +60,30 @@ export const testComponentSnapshotsWithFixtures = (Component, fixtures) =>
+ * @param  {Number} states the depth of dispatch calls
+ * @return calls result tree to the given depth - array for each branch of calls
+ */
+export const runActionInDepth = (runAction, depth = 1) =>
```
@amirfefer what do you think about his one? I have figured, I would like to test deeper caller tree of the dispatch, not just the first level, so I wrote this helper for it.

amirfefer replied:
@sharvit , any thoughts? could be added to `testHelpers` repo

amirfefer replied:
@sharvit ping

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/__tests__/TemplateGeneratorActions.test.js:
```diff
@@ -0,0 +1,83 @@
+        .mockImplementationOnce(async () => noContentResponse)
+        .mockImplementationOnce(async () => generatedReportResponse);
+
+      runActionInDepth(() => actions.generateTemplate(), 3).then(callTree => {
```
Here is example usage of the new test helper.

amirfefer replied:
wouldn't `testActionSnapshotWithFixtures()` test the same with a snapshot?

ezr-ondrej replied:
It would not, because it is generating calls only in depth of 1, so if dispatch is called with another dispatch function, and from that again, the snapshot won't have that covered, just will state function in the tree. Another thing is, that this is more explicit, than just a snapshot, I prefer to be more explicit in the tests (otherwise we can update the testActionSnapshotFixtures to take the depth argument, but not in this PR i guess)

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGenerator.js:
```diff
@@ -0,0 +1,39 @@
+const infoMsg =
+  'This will generate report %s. Based on its definition, it can take a long time to process.';
+const pollingMsg =
+  "Report %s is now being generated, the download will start once it's done. In case you don't want to wait, use the following link to come later on for the result.";
```
Not sure about the wording though :thinking:

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -129,12 +170,31 @@ def action_permission
+        if User.current.admin? || options['user_id'].to_i == User.current.id
+          @composer = ReportComposer.new(composer_attrs)
+        else
+          deny_access(_('Report data are available only for the user who started the report'))
```
`Data are available only for the user who triggered the report and administrators`

ares commented on app/jobs/stored_values_cleanup_job.rb:
```diff
@@ -0,0 +1,18 @@
+class StoredValuesCleanupJob < ApplicationJob
+  def perform(options = {})
+    StoredValue.expired(options[:ago]).delete_all
```
while this is faster, would it be better to use `destroy_all` to run all callbacks if someone adds some in future?

ezr-ondrej replied:
Yeah, it would so I did it.

ares commented on app/models/stored_value.rb:
```diff
@@ -0,0 +1,22 @@
+# This class represents temporary storage for values processed in background.
```
This comment is already obsolete, can we just drop it? Also for the one below, roadmap or generic plans about future should be captured in redmine or discourse, not in code comments :-)

ezr-ondrej replied:
Dropped.

ares commented on app/models/stored_value.rb:
```diff
@@ -0,0 +1,22 @@
+# This is a database solution, wich should not be the only one, as it has many limitations.
+class StoredValue < ApplicationRecord
+  scope :valid, lambda {
+    f = arel_table[:expire_at]
```
I'd prefer using active record here

```ruby
scope :expired, lambda {
    where([ "expire_at < ?", Time.now ]).or(:expire_at => nil)
}
```

seems more readable while should still work, the same for expired below

ezr-ondrej replied:
I like arel more, but I agree, this is more readable, so in this case works better, done that.

ares reviewed (changes requested):
Left few small comments. From testing:

1) seems like inputs are not passed to the job, when I click on submit, nothing happens, this is in my log

```
2019-03-05T17:15:08 [E|app|0f69f567] Unprocessable entity ReportComposer (id: new):
 |   Input Host filter: Value can't be blank
 | 
2019-03-05T17:15:08 [I|app|0f69f567]   Rendering api/v2/errors/unprocessable_entity.json.rabl within api/v2/layouts/error_layout
2019-03-05T17:15:08 [I|app|0f69f567]   Rendered api/v2/errors/unprocessable_entity.json.rabl within api/v2/layouts/error_layout (5.6ms)
2019-03-05T17:15:08 [D|app|0f69f567] Body: {
 |   "error": {"errors":{"base":["Input Host filter: Value can't be blank"]},"full_messages":["Input Host filter: Value can't be blank"]}
 | }
 | 
2019-03-05T17:15:08 [I|app|0f69f567] Completed 422 Unprocessable Entity in 380ms (Views: 13.9ms | ActiveRecord: 19.1ms)
```

through API btw it works fine

2) if the template rendering fails, e.g. safemode does not allow rendering, I still don't see anything in UI, just the trace in production.log

3) while the template is rendered (if there's no error), this is what I see
![report_progess](https://user-images.githubusercontent.com/109773/53819741-cb376800-3f6a-11e9-81b8-2f1c7a9236e4.png)

I think it would be nicer to either generate the link with full url. Or move it to the warning and make "following link" to be clickable. The relative link where most of it is ugly UUID does not look too nice.

This page should also say that the link will be available for $x hours after it's ready. And there should be some spinner icon so people know it's doing something on background.

4) The cleanup task is stopped warning, the error is 

```
TypeError: can't convert nil into an exact number
Backtrace:
/home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/core_ext/time/calculations.rb:275:in `-'
/home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/core_ext/time/calculations.rb:275:in `minus_with_duration'
/home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/core_ext/time/calculations.rb:286:in `minus_with_coercion'
/home/ares/Projekty/Zdrojaky/foreman/app/models/stored_value.rb:9:in `block in <class:StoredValue>'
```

5) every template rendering create task with a name `TemplateRenderJob`, could it have a nicer name, such as "Report Template #{x} rendering". If the dynamic part is problematic, at least "Report Template Rendering" please

6) Fetching the scheduled report from UI through API works but is not great, because if the content is not ready yet, users ends up with empty white page. If user is not authorized, they end up with white page with ugly json wrapped error message. In worst case we can add that in separate PR, but it would be great to have download page, that would inform me that the report is still being processed and would start polling when I enter, or tell me I don't have permission while I still have layout so I can navigate to other page.

The happy path though works nice, we're getting closer :-)

ezr-ondrej committed:
Fixes #26020 - report templates in background, UI improvements

ezr-ondrej commented:
I have:

- Rewrote the GUI to have its own page, so the user can return to that page whenever he wants.
- Rewrote the javascript a bit to fit the new page - it is rendering the link as a button and only when ready (since the page you are going to return to is the one you are on).

ares commented on package.json:
```diff
@@ -75,6 +75,7 @@
     "datatables.net": "~1.10.12",
     "datatables.net-bs": "~1.10.12",
     "diff": "~3.0.0",
+    "file-saver": "^2.0.1",
```
could we get ack from @theforeman/packaging team? (cc @ekohl explicitly as I'm not sure if this is the correct handle :-)

amirfefer replied:
you can use axios directly for downloading content: https://gist.github.com/javilobo8/097c30a233786be52070986d8cdb1743

ezr-ondrej replied:
I know, but it is messing with the DOM (adding a link and clicking it, sounds like not really clean way to do things).
But if you don't like this one, we already have the download link (for automatic dowload failures), we can use that one. directly and let the browser download the file.

ekohl commented on package.json:
```diff
@@ -75,6 +75,7 @@
     "datatables.net": "~1.10.12",
     "datatables.net-bs": "~1.10.12",
     "diff": "~3.0.0",
+    "file-saver": "^2.0.1",
```
https://github.com/theforeman/foreman-packaging/pull/3520

ezr-ondrej replied:
Thanks a lot! :+1:

ekohl reviewed (dismissed):
:+1: from a packaging perspective.

ares commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGenerator.js:
```diff
@@ -0,0 +1,79 @@
+
+const pollingMsg = `
+  Report %s is now being generated, the download will start once it's done.
+  You can come to this site anytime to get the results.
```
You can come to this site anytime to get the results.s/You can come to this page later to get the results. The is stored for 24 hours./

ares reviewed (changes requested):
The UI now works much better. While testing I found few more small things.

1) When I enter the `/templates/report_templates/758-sleep/report_data?job_id=` with wrong job id, I'm seeing ugly error `caused by Sequel::DatabaseError: PG::InvalidTextRepresentation: ERROR: invalid input syntax for type uuid: "x" LINE 1: ... * FROM "dynflow_execution_plans" WHERE ("uuid" = 'x') LIMIT... ^`. It should say such report was not found. That can happen if user does not copy the whole value. If I enter already expired record, I'm getting also quite ugly `searching: execution_plan by: {:uuid=>"c1f6abad-8d42-46af-ac7b-5fbe3b16eba1"}`.

2) the task name in Monitor -> Tasks is now "undefined method `[]' for nil:NilClass (NoMethodError) ", meaning the point 5 from previous list is not addressed

including this trace:

```/home/ares/Projekty/Zdrojaky/foreman/app/jobs/template_render_job.rb:14:in `template_id' /home/ares/Projekty/Zdrojaky/foreman/app/jobs/template_render_job.rb:18:in `humanized_name' /home/ares/Projekty/Zdrojaky/foreman-tasks/app/models/foreman_tasks/task/dynflow_task.rb:165:in `get_humanized' /home/ares/Projekty/Zdrojaky/foreman-tasks/app/models/foreman_tasks/task.rb:256:in `to_label' /home/ares/Projekty/Zdrojaky/foreman-tasks/app/models/foreman_tasks/task/dynflow_task.rb:219:in `set_action_field' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/callbacks.rb:426:in `block in make_lambda' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/callbacks.rb:247:in `block in halting' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/callbacks.rb:517:in `block in invoke_after' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/callbacks.rb:517:in `each' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/callbacks.rb:517:in `invoke_after' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/callbacks.rb:133:in `run_callbacks' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.2.1/lib/active_support/callbacks.rb:816:in `_run_validation_callbacks' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activemodel-5.2.1/lib/active_model/validations/callbacks.rb:118:in `run_validations!' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activemodel-5.2.1/lib/active_model/validations.rb:339:in `valid?' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/validations.rb:67:in `valid?' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/validations.rb:84:in `perform_validations' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/validations.rb:52:in `save!' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/transactions.rb:315:in `block in save!' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/transactions.rb:387:in `block in with_transaction_returning_status' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/database_statements.rb:254:in `block in transaction' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/transaction.rb:239:in `block in within_new_transaction' /home/ares/.rbenv/versions/2.4.0/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/transaction.rb:236:in `within_new_transaction' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/database_statements.rb:254:in `transaction' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/transactions.rb:212:in `transaction' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/transactions.rb:385:in `with_transaction_returning_status' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/transactions.rb:315:in `save!' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activerecord-5.2.1/lib/active_record/suppressor.rb:48:in `save!' /home/ares/Projekty/Zdrojaky/foreman-tasks/app/models/foreman_tasks/task/dynflow_task.rb:24:in `update_from_dynflow' /home/ares/Projekty/Zdrojaky/foreman-tasks/lib/foreman_tasks/dynflow/persistence.rb:43:in `on_execution_plan_save' /home/ares/Projekty/Zdrojaky/foreman-tasks/lib/foreman_tasks/dynflow/persistence.rb:12:in `block in save_execution_plan' /home/ares/Projekty/Zdrojaky/foreman-tasks/lib/foreman_tasks/dynflow/persistence.rb:10:in `tap' /home/ares/Projekty/Zdrojaky/foreman-tasks/lib/foreman_tasks/dynflow/persistence.rb:10:in `save_execution_plan' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/persistence.rb:73:in `save_execution_plan' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/execution_plan.rb:432:in `save' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/execution_plan.rb:137:in `update_state' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/director.rb:224:in `update_execution_plan_state' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/director.rb:160:in `finish_manager' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/director.rb:149:in `unless_done' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/director.rb:112:in `work_finished' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/executors/parallel/core.rb:49:in `work_finished' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/actor.rb:6:in `on_message' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/executors/parallel/core.rb:109:in `on_message' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/context.rb:46:in `on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/executes_context.rb:7:in `on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/dynflow-1.2.2/lib/dynflow/actor.rb:26:in `on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/awaits.rb:15:in `on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/sets_results.rb:14:in `on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/buffer.rb:38:in `process_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/buffer.rb:31:in `process_envelopes?' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/buffer.rb:20:in `on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/termination.rb:55:in `on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/removes_child.rb:10:in `on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/abstract.rb:25:in `pass' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/behaviour/sets_results.rb:14:in `on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/core.rb:162:in `process_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/core.rb:96:in `block in on_envelope' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/core.rb:119:in `block (2 levels) in schedule_execution' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/synchronization/mutex_lockable_object.rb:41:in `block in synchronize' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/synchronization/mutex_lockable_object.rb:41:in `synchronize' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/synchronization/mutex_lockable_object.rb:41:in `synchronize' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-edge-0.4.1/lib-edge/concurrent/actor/core.rb:116:in `block in schedule_execution' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/executor/serialized_execution.rb:18:in `call' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/executor/serialized_execution.rb:96:in `work' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/executor/serialized_execution.rb:77:in `block in call_job' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/executor/ruby_thread_pool_executor.rb:348:in `run_task' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/executor/ruby_thread_pool_executor.rb:337:in `block (3 levels) in create_worker' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `loop' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `block (2 levels) in create_worker' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `catch' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/concurrent-ruby-1.1.4/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `block in create_worker' /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

ares commented:
@amirfefer could you please give it a look for frontend part? I can imagine more usability tweaks but they can be added later in separate PR.

ezr-ondrej committed:
Fixes #26020 - report templates generation tweeks

ezr-ondrej commented:
1. addressed by last commit - see the `rescue nil` - I dont like consuming all exceptions like that, but we would have to rescue Key error from dynflow (non existent task, valid uuid) and validate uuid, which would be ugly RegExp :disappointed: 
2. is an issue in foreman-tasks introduced with dynflow 1.22 (see https://github.com/theforeman/foreman-tasks/pull/396), and obviously my code not being ready for it, so now it is returning the basic `Render report template` till the PR is merged, after it should give the specific `Render report %s`

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -187,7 +189,7 @@ def action_permission
       def load_and_authorize_plan
         @plan = load_dynflow_plan(params[:job_id])
-        return not_found(_('Report schedule not found, please ensure you have got right job_id')) if @plan.nil?
+        return not_found(_('Report not found, please ensure you have got right job_id')) if @plan.nil?
```
`Report not found, please ensure you used the correct job_id` would I think sound better

ezr-ondrej replied:
Done :)

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -198,6 +200,8 @@ def load_and_authorize_plan
 
       def load_dynflow_plan(plan_id)
         Rails.application.dynflow.world.persistence.load_execution_plan(plan_id)
+      rescue
```
should we at least log the error in debug mode? there's an exception logger I think

ezr-ondrej replied:
:dart: that is what I was looking for, thanks :+1:

ares commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGenerator.js:
```diff
@@ -0,0 +1,79 @@
+  You can come to this page later to get the results. The result is available for 24 hours.
+`;
+const doneMsg =
+  'Generation of report %s has completed. Download should start automatically. In case it does not, please use download button below.';
```
`Generating of the report %s has been completed. Download should start automatically. In case it does not, please use the download button below.`

ezr-ondrej replied:
Yeah sounds better! :+1:

ares reviewed (changes requested):
We did a quick review with @terezanovotna and found few strings that would be good to improve. She was OK with the pages content. When we display the banner that report is being generated, we should activate the spinner next to locations selector. Once we're done, we should deactive it. Also if I click download button it activates the spinner, we need to disable it again.

One idea for them, that is out of the scope for this PR, is to use empty state component. That would probably look nicer. Also it would be absolutely great to have a notification in drawer when the report is generated with the link to download, again not in this PR and I'm happy to open the RFE. Last RFE we identified is that the "Submit" button should be labeled "Generate the report", which is again out of the scope since it already exists, but I would accept it here too :-)

If you could fix the strings, I'll wait for feedback from @amirfefer, but IMHO it's good to be merged afterwards.

ares commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/__tests__/TemplateGenerator.test.js:
```diff
@@ -0,0 +1,35 @@
+    polling: false,
+    generationError: '422 unprocessable entity',
+    generationErrorMessages: [
+      { message: 'Eh there was no method error during the render :(' },
```
if we change generation tome something else, it needs to be modified here too

ezr-ondrej replied:
:dart:

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -121,6 +123,55 @@ def generate
         render_error 'standard_error', :status => :internal_error, :locals => { :exception => e }
       end
 
+      api :POST, "/report_templates/:id/schedule_report/", N_("Schedule a report generation")
```
s/Schedule a report generation/Schedule generating of a report/ sound better IMO

ezr-ondrej replied:
:+1:

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -121,6 +123,55 @@ def generate
+      param :input_values, Hash, :desc => N_('Hash of input values where key is the name of input, value is the value for this input')
+      param :gzip, :bool, desc: N_('Compress the report uzing gzip')
+      returns :code => 200, :desc => "a successful response" do
+        property :job_id, String, :desc => "An id of generation job, pass it to the report_data action for data retrieval."
```
"An id of generation job, pass it to the report_data action for data retrieval."
probably sounds better as
"A job ID which generates the report. To be used with report_data API endpoint for report data retrieval"

ezr-ondrej replied:
I've come up with "An ID of job, which generates report. To be used with report_data API endpoint for report data retrieval." will that be ok?

ares replied:
It's good, probably there should be `the` before `report` in the first sentence.

ares commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -121,6 +123,55 @@ def generate
+      POST /api/report_templates/:id/schedule_report/
+      200
+      {
+        "job_id": UNIQUE-REPORT-GENERATION-JOB-ID
```
s/GENERATION/GENERATING/ here and on line below

ezr-ondrej replied:
:+1:

ares commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGenerator.js:
```diff
@@ -0,0 +1,82 @@
+  </Alert>
+);
+
+class TemplateGenerator extends React.Component {
```
s/generation/generating/ in the whole file would probably be better or perhaps `generateError`?

ezr-ondrej replied:
Yeah after a while I realized what is the point (generation coul have different meaning). That case generating sounds better to me. generateError sounds like we are generating the error, and the idea is not to do that if possible, right? :laughing:

ares commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGeneratorReducer.js:
```diff
@@ -0,0 +1,41 @@
+      return state.merge({
+        scheduleInProgress: false,
+        polling: false,
+        generationError: payload.error.message,
```
generation -> generating here as well

ezr-ondrej replied:
:+1:

ares commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/index.js:
```diff
@@ -0,0 +1,30 @@
+  pick(templates, [
+    'polling',
+    'dataUrl',
+    'generationError',
```
also to be changed

ezr-ondrej replied:
:+1:

ares reviewed (changes requested):
Few more strings adjustments please, given we're waiting on frontend code ack :-)

ezr-ondrej commented:
I've done the renaming and forced the download to be in new tab (I haven't figure out how to force the turbolinks not to kick-in). The others would make the PR even complicated and I would prefer follow-up PR, if it wouldn't be a blocker for you.

ares commented:
Thanks, one nit to consider. I'll retest once more after frontend ack, but I think it's acceptable to have further UI improvements in follow up PR :+1:

amirfefer commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGenerator.js:
```diff
@@ -0,0 +1,80 @@
+import React from 'react';
```
`templates` is a big part , what about make `Templates` dir with subdir of `components `?
`Template/components/TemplateGenerator/TemplateGenerator.js`

ezr-ondrej replied:
even if TemplateGenerator is a high level component and there is no Template component?
wouldn't be `templates/TemplateGenerator/TemplateGenerator.js` more reflective of the fact of being high level component?

amirfefer replied:
I'd like to keep consistency with #6531

ezr-ondrej replied:
But I don't see the pattern you are describing used there :thinking: don't you mean to make it `ReportTemplate/ReportTemplateGenerator.js` ? That I would understand, and agreed, but I don't feel right about putting it deeper.

amirfefer commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGenerator.js:
```diff
@@ -0,0 +1,80 @@
+  getError() {
+    const { generatingError, generatingErrorMessages } = this.props;
+    return (
+      (generatingErrorMessages &&
```
nit: I prefer make it more simple when logic operands mixed together - readability gets improved  
something like:
```js
const errors = generatingErrorMessages.length  &&
    generatingErrorMessages.map(e => e.message).join('\n')

return (errors || generatingError) 
```

ezr-ondrej replied:
yeah, feels better, done that.

amirfefer commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGenerator.js:
```diff
@@ -0,0 +1,80 @@
+  polling: PropTypes.bool,
+  dataUrl: PropTypes.string,
+  generatingError: PropTypes.string,
+  generatingErrorMessages: PropTypes.array,
```
could you use `arrayOf` or `shape` instead of just `array?`

ezr-ondrej replied:
:dart: thanks :+1:

amirfefer commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGeneratorActions.js:
```diff
@@ -0,0 +1,67 @@
+
+const defaultPollingInterval = 2500;
+const pollingInterval =
+  process.env.NOTIFICATIONS_POLLING || defaultPollingInterval;
```
do we wish using the same polling as notifications for templates?
I can think of cases that users would use different polling interval for notifications and reports

ezr-ondrej replied:
There is a discussion with ohad wabout that above. I wouldn't bring special variable for this one, as it is only if you choose to stay on one specific page. So if not the same (user can choose to prolong the notifications polling, but still wants to poll frequently for reports), I would just come up with reasonable default and wouldn't let users to change it.

ezr-ondrej replied:
I have just went with the default of 3secs.

amirfefer commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGeneratorActions.js:
```diff
@@ -0,0 +1,67 @@
+  return [errorResponse.data];
+};
+
+export const pollReportData = pollUrl => dispatch => {
```
we have other actions for polling, I wonder about creating a general action for polling requests

ezr-ondrej replied:
I guess we've got just notifications.
Anyway I agree we should, but I would prefer to do that as a followup, as I don't feel it like a blocker, but if it is for you, I can do that.

amirfefer replied:
In subscriptions page we also use the same polling mechanism (which caused some annoying bugs),  it's not a blocker, but I concern that it would be missed.
Having a middlewere for networking is important, especially when we got complex sceanrio like polling.

 @sharvit has refactored the polling logic in katello, but it haven't merged yet, if you have time you can reuse it (https://github.com/sharvit/katello/commit/967fceae2cf2fb9db5870251a17a3a3360710d32#diff-b9290ce729f3b4b0a868fb93e92be9fb)

amirfefer commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGeneratorReducer.js:
```diff
@@ -0,0 +1,41 @@
+export default (state = initialState, { type, payload }) => {
+  switch (type) {
+    case TEMPLATE_GENERATE_REQUEST:
+      return state.merge({ scheduleInProgress: true });
```
it's only one value, how about using `set` here?`

ezr-ondrej replied:
:dart: thanks

amirfefer commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/index.js:
```diff
@@ -0,0 +1,30 @@
+
+// map state to props
+const mapStateToProps = ({ templates }) =>
+  pick(templates, [
```
could you use selectors ? 
afaik `pick` will create a new object, so even if attributes haven't been changed, a re-render will occur

ezr-ondrej replied:
I haven't realized that about `pick`. I have created selectors, they are simple ones, but you are right, feels better.

amirfefer reviewed (changes requested):
Thanks, @ezr-ondrej !
great work :+1: 

just few comments about js code.

ezr-ondrej committed:
Fixes #26020 - report templates React tweeks

ezr-ondrej commented:
@amirfefer thanks for the review, I have addressed your comments, please take another look when convenient, thanks!

amirfefer commented on webpack/assets/javascripts/react_app/components/TemplateGenerator/TemplateGeneratorSelectors.js:
```diff
@@ -0,0 +1,15 @@
+const generatingErrorMessages = state =>
+  selectTemplates(state).generatingErrorMessages;
+
+export const selectGeneratingProps = createStructuredSelector({
```
using reselct here is redundant because no calculations have made here, so memorization won't be effective.
simple selectors would be enough.
Plus. could you add a test? we have a test helper for selectors as well.

ezr-ondrej replied:
:dart: thanks, I thought the `createStructuredSelector` is meant to take out the boilerplate of combined selector, but it does memorization as well.
I believe I don't need the helper here as my selector is really simple and I believe my test is as well, I have added one.

amirfefer commented on app/views/report_templates/report_data.html.erb:
```diff
@@ -0,0 +1,8 @@
+<% title _("Download generated report") %>
```
I wonder if having a full page for downloading process is needed? looks a bit odd
how about using the generator page itself, or showing a modal during the process, likewise importing manifest in katello

ezr-ondrej replied:
I have done just that previously, but it has to control the page outside of it and that doesn't feel right. I hope `generate` page to be react page in the feature and this component to be one of the components there.

amirfefer reviewed (changes requested):
Thanks @ezr-ondrej  :+1: 
I left few comments (mostly nitpciks)

amirfefer reviewed (dismissed):
Thanks @ezr-ondrej :+1: 
js part looks great, LGTM

ares commented:
@ezr-ondrej so we have 2 acks now, could you please rebase to new develop in order to get tests running? once Mr. Jenkins confirms, I'm happy to merge.

ezr-ondrej committed:
Fixes #26020 - english nits

ezr-ondrej commented:
[test foreman]

ares commented:
Thanks @ezr-ondrej, great patch. Merging!
